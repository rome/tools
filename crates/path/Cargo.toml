[package]
name = "path"
version = "0.0.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
core = { path = "../core" }

[dev-dependencies]
