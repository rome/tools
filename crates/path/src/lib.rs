//! This module is responsible to manage paths inside Rome.
//! It is a small wrapper around [path::PathBuf] but it is also able to
//! give additional information around the the file that holds:
//! - the [FileHandlers] for the specific file
//! - shortcuts to open/write to the file
use core::{file_handlers::ExtensionHandler, App};
use std::{fs::File, io::Write, ops::Deref, path::PathBuf};

pub struct RomePath<'handler> {
	file: PathBuf,

	handler: Option<&'handler Box<dyn ExtensionHandler>>,
}

impl<'handler> Deref for RomePath<'handler> {
	type Target = PathBuf;

	fn deref(&self) -> &Self::Target {
		&self.file
	}
}

impl<'handler> RomePath<'handler> {
	pub fn new(path_to_file: &str) -> Self {
		Self {
			file: PathBuf::from(path_to_file),
			handler: None,
		}
	}

	/// Deduce the file handler based on the extension of the file.
	///
	/// Any error will default to the base file handler for now.
	///
	///
	/// ```rust
	/// use path::RomePath;
	/// use core::create_app;
	/// let app = create_app();
	/// let file = RomePath::new("file.js").deduce_handler(&app);
	/// let handler = file.get_handler();
	/// let expected = app.get_js_handler("js").unwrap();
	/// assert_eq!(handler.unwrap(), expected)
	/// ```
	pub fn deduce_handler(mut self, app: &'handler App) -> Self {
		if let None = self.extension() {
			return self;
		}
		let extension = self.extension().unwrap().to_str().unwrap();

		if let Some(handler) = app.get_handler(extension) {
			self.handler = Some(handler);
		}

		self
	}

	// TODO: handle error with diagnostic?
	/// Opens a file and returns a [File] in write mode
	pub fn open(&self) -> File {
		File::open(&self.file).expect("cannot open the file to format")
	}


	/// Accepts a file opened in read mode and saves into it
	pub fn save(&mut self, content: &str) -> Result<(), std::io::Error> {
		let mut file_to_write = File::create(&self.file).unwrap();
		file_to_write.write_all(content.as_bytes())
	}

	/// Returns the current handler associated to the file.
	///
	/// You need to call [deduce_handler] first in order to receive one. If not, [None] is always returned.
	pub fn get_handler(&self) -> Option<&Box<dyn ExtensionHandler>> {
		self.handler
	}
}

#[cfg(test)]
mod test {
	use crate::RomePath;
	use core::{
		create_app,
		file_handlers::{javascript::JsFileHandler, ExtensionHandler},
	};

	#[test]
	fn deduce_handler() {
		let app = create_app();
		let file = RomePath::new("file.js").deduce_handler(&app);
		let handler = file.get_handler();
		let expected = JsFileHandler {};
		assert_eq!(
			handler.unwrap().capabilities().format,
			expected.capabilities().format
		);
		assert_eq!(
			handler.unwrap().capabilities().lint,
			expected.capabilities().lint
		)
	}
}
