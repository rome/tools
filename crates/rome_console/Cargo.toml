[package]
authors.workspace    = true
description          = "Expose utilities to print HTML-syntax messages"
documentation        = "https://rustdocs.rome.tools/rome_console/index.html"
edition.workspace    = true
license.workspace    = true
name                 = "rome_console"
repository.workspace = true
version              = "0.0.1"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
atty           = "0.2.14"
rome_markup    = { workspace = true }
rome_text_size = { workspace = true }
schemars       = { workspace = true, optional = true }
serde          = { workspace = true, optional = true, features = ["derive"] }
termcolor      = "1.1.2"
unicode-width  = "0.1.9"

[dev-dependencies]
trybuild = "1.0"

[features]
serde_markup = ["serde", "schemars"]
