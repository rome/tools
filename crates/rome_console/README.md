# `rome_console`

The crate contains a general abstraction over printing messages (formatted with markup) and diagnostics to a console.

## Local installation

```toml
rome_console = { version = "0.0.0", path = "../rome_console" }
```

## Usage example

The `Console` trait can be used to print two types of information to the user: messages (in the form of markup) and diagnostics:

```rust
console.message(markup! {
    <Info>"Processed "<Emphasis>{count}</Emphasis>" files"</Info>
});

console.diagnostic(
    &mut files,
    Diagnostics::error(file_id, code, title),
);
```

The following markup elements are supported:
- `Emphasis`: Print the content in bold text
- `Dim`: Print the content in dimmed text
- `Italic`: Print the content in italic text
- `Underline`: Print the content in underlined text
- `Error`: Set the text color to red
- `Success`: Set the text color to green
- `Warn`: Set the text color to yellow
- `Info`: Set the text color to blue

*Note*: Markup elements that change the "font" of the printed text (`Emphasis`, `Dim`, `Italic` and `Underline`) are not supported by the native Windows Console API and will instead get printed as ANSI control codes if the current terminal supports it, or will be ignored entirely
