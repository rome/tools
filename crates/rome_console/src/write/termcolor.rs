use std::{
    fmt::{self, Write as _},
    io,
};

use termcolor::{Color, ColorSpec, WriteColor};
use unicode_width::UnicodeWidthChar;

use crate::{fmt::MarkupElements, MarkupElement};

use super::Write;

/// Adapter struct implementing [Write] over types implementing [WriteColor]
pub struct Termcolor<W>(pub W);

impl<W> Write for Termcolor<W>
where
    W: WriteColor,
{
    fn write_str(&mut self, elements: &MarkupElements, content: &str) -> io::Result<()> {
        with_format(&mut self.0, elements, |writer| {
            let mut adapter = SanitizeAdapter {
                writer,
                error: Ok(()),
            };

            match adapter.write_str(content) {
                Ok(()) => Ok(()),
                Err(..) => {
                    if adapter.error.is_err() {
                        adapter.error
                    } else {
                        // SanitizeAdapter can only fail if the underlying
                        // writer returns an error
                        unreachable!()
                    }
                }
            }
        })
    }

    fn write_fmt(&mut self, elements: &MarkupElements, content: fmt::Arguments) -> io::Result<()> {
        with_format(&mut self.0, elements, |writer| {
            let mut adapter = SanitizeAdapter {
                writer,
                error: Ok(()),
            };

            match adapter.write_fmt(content) {
                Ok(()) => Ok(()),
                Err(..) => {
                    if adapter.error.is_err() {
                        adapter.error
                    } else {
                        Err(io::Error::new(
                            io::ErrorKind::Other,
                            "a Display formatter returned an error",
                        ))
                    }
                }
            }
        })
    }
}

/// Applies the current format in `state` to `writer`, calls `func` to
/// print a piece of text, then reset the printing format
fn with_format<W>(
    writer: &mut W,
    state: &MarkupElements,
    func: impl FnOnce(&mut W) -> io::Result<()>,
) -> io::Result<()>
where
    W: WriteColor,
{
    let mut color = ColorSpec::new();
    let mut link = None;
    let mut inverse = false;

    state.for_each(&mut |elements| {
        for element in elements {
            match element {
                MarkupElement::Inverse => {
                    inverse = !inverse;
                }
                MarkupElement::Hyperlink { href } => {
                    link = Some(href);
                }
                _ => {
                    element.update_color(&mut color);
                }
            }
        }

        Ok(())
    })?;

    if inverse {
        let fg = color.fg().map_or(Color::White, |c| *c);
        let bg = color.bg().map_or(Color::Black, |c| *c);
        color.set_bg(Some(fg));
        color.set_fg(Some(bg));
    }

    if let Err(err) = writer.set_color(&color) {
        writer.reset()?;
        return Err(err);
    }

    let mut reset_link = false;
    if let Some(href) = link {
        // `is_synchronous` is used to check if the underlying writer
        // is using the Windows Console API, that does not support ANSI
        // escape codes. Generally this would only be true when running
        // in the legacy `cmd.exe` terminal emulator, since in modern
        // clients like the Windows Terminal ANSI is used instead
        if writer.supports_color() && !writer.is_synchronous() {
            write!(writer, "\x1b]8;;{href}\x1b\\")?;
            reset_link = true;
        }
    }

    let result = func(writer);

    if reset_link {
        write!(writer, "\x1b]8;;\x1b\\")?;
    }

    writer.reset()?;
    result
}

/// Adapter [fmt::Write] calls to [io::Write] with sanitization,
/// implemented as an internal struct to avoid exposing [fmt::Write] on
/// [Termcolor]
struct SanitizeAdapter<W> {
    writer: W,
    error: io::Result<()>,
}

impl<W> fmt::Write for SanitizeAdapter<W>
where
    W: WriteColor,
{
    fn write_str(&mut self, content: &str) -> fmt::Result {
        let mut buffer = [0; 4];

        for item in content.chars() {
            // Replace non-whitespace, zero-width characters with the Unicode replacement character
            let is_whitespace = item.is_whitespace();
            let is_zero_width = UnicodeWidthChar::width(item).map_or(true, |width| width == 0);
            let item = if !is_whitespace && is_zero_width {
                char::REPLACEMENT_CHARACTER
            } else if cfg!(windows) || !self.writer.supports_color() {
                // Unicode is currently poorly supported on most Windows
                // terminal clients, so we always strip emojis in Windows
                unicode_to_ascii(item)
            } else {
                item
            };

            item.encode_utf8(&mut buffer);
            if let Err(err) = self.writer.write_all(&buffer[..item.len_utf8()]) {
                self.error = Err(err);
                return Err(fmt::Error);
            }
        }

        Ok(())
    }
}

/// Replace emoji characters with similar but more widely supported ASCII
/// characters
fn unicode_to_ascii(c: char) -> char {
    match c {
        '\u{2714}' => '\u{221a}',
        '\u{2139}' => 'i',
        '\u{26a0}' => '!',
        '\u{2716}' => '\u{00d7}',
        _ => c,
    }
}

#[cfg(test)]
mod tests {
    use std::{fmt::Write, str::from_utf8};

    use rome_markup::markup;
    use termcolor::Ansi;

    use crate as rome_console;
    use crate::fmt::Formatter;

    use super::{SanitizeAdapter, Termcolor};

    #[test]
    fn test_sanitize() {
        // Sanitization should leave whitespace control characters (space,
        // tabs, newline, ...) and non-ASCII unicode characters as-is but
        // redact zero-width characters (RTL override, null character, bell,
        // zero-width space, ...)
        const INPUT: &str = "t\tes t\r\n\u{202D}t\0es\x07t\u{202E}\nt\u{200B}esüêõt";
        const OUTPUT: &str = "t\tes t\r\n\u{FFFD}t\u{FFFD}es\u{FFFD}t\u{FFFD}\nt\u{FFFD}esüêõt";

        let mut buffer = Vec::new();

        {
            let writer = termcolor::Ansi::new(&mut buffer);
            let mut adapter = SanitizeAdapter {
                writer,
                error: Ok(()),
            };

            adapter.write_str(INPUT).unwrap();
            adapter.error.unwrap();
        }

        assert_eq!(from_utf8(&buffer).unwrap(), OUTPUT);
    }

    #[test]
    fn test_hyperlink() {
        const OUTPUT: &str = "\x1b[0m\x1b]8;;https://rome.tools/\x1b\\link\x1b]8;;\x1b\\\x1b[0m";

        let mut buffer = Vec::new();
        let mut writer = Termcolor(Ansi::new(&mut buffer));
        let mut formatter = Formatter::new(&mut writer);

        formatter
            .write_markup(markup! {
                <Hyperlink href="https://rome.tools/">"link"</Hyperlink>
            })
            .unwrap();

        assert_eq!(from_utf8(&buffer).unwrap(), OUTPUT);
    }
}
