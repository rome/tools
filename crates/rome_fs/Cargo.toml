[package]
name = "rome_fs"
version = "0.0.0"
edition = "2021"
authors = ["Rome Tools Developers and Contributors"]
repository = "https://github.com/rome/tools"
license = "MIT"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rome_diagnostics = { path = "../rome_diagnostics" }
indexmap = "1.8.0"
parking_lot = { version = "0.12.0", features = ["arc_lock"] }
rayon = "1.5.1"
crossbeam = "0.8.1"
tracing = { version = "0.1.31", default-features = false, features = ["std"] }
serde = { version = "1.0.136", features = ["derive"], optional = true }
schemars = { version = "0.8.10", optional = true }

[dev-dependencies]

[features]
serde = ["dep:serde", "schemars"]
