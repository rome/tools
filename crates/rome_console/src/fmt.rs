use std::{borrow::Cow, fmt, io, time::Duration};

use termcolor::{ColorSpec, WriteColor};

use crate::{markup, Markup, MarkupElement};

/// A stack-allocated linked-list of [MarkupElement] slices
enum MarkupElements<'a> {
    Root,
    Node(&'a Self, &'a [MarkupElement]),
}

impl<'a> MarkupElements<'a> {
    /// Iterates on all the element slices depth-first
    fn for_each(&self, func: &mut impl FnMut(&'a [MarkupElement])) {
        if let Self::Node(parent, elem) = self {
            parent.for_each(func);
            func(elem);
        }
    }
}

/// The [Formatter] is the `rome_console` equivalent to [std::fmt::Formatter]:
/// it's never constructed directly by consumers, and can only be used through
/// the mutable reference passed to implementations of the [Display] trait).
/// It manages the state of the markup to print, and implementations of
/// [Display] can call into its methods to append content into the current
/// printing session
pub struct Formatter<'fmt> {
    /// Stack of markup elements currently applied to the text being printed
    state: MarkupElements<'fmt>,
    /// Inner IO writer this [Formatter] will print text into
    writer: &'fmt mut dyn WriteColor,
}

impl<'fmt> Formatter<'fmt> {
    /// Create a new instance of the [Formatter] using the provided `writer` for printing
    pub(crate) fn new(writer: &'fmt mut dyn WriteColor) -> Self {
        Self {
            state: MarkupElements::Root,
            writer,
        }
    }

    /// Return a new instance of the [Formatter] with `elements` appended to its element stack
    fn with_elements<'b>(&'b mut self, elements: &'b [MarkupElement]) -> Formatter<'b> {
        Formatter {
            state: MarkupElements::Node(&self.state, elements),
            writer: self.writer,
        }
    }

    /// Write a piece of markup into this formatter
    pub fn write_markup(&mut self, markup: Markup) -> io::Result<()> {
        for node in markup.0 {
            let mut fmt = self.with_elements(node.elements);
            node.content.fmt(&mut fmt)?;
        }

        Ok(())
    }

    /// Applies the current format in `state` to `writer`, calls `func` to
    /// print a piece of text, then reset the printing format
    fn with_format(
        &mut self,
        func: impl FnOnce(&mut dyn WriteColor) -> io::Result<()>,
    ) -> io::Result<()> {
        let mut color = ColorSpec::new();
        self.state.for_each(&mut |elements| {
            for element in elements {
                element.update_color(&mut color);
            }
        });

        if let Err(err) = self.writer.set_color(&color) {
            self.writer.reset()?;
            return Err(err);
        }

        let result = func(self.writer);
        self.writer.reset()?;
        result
    }

    /// Write a slice of text into this formatter
    pub fn write_str(&mut self, content: &str) -> io::Result<()> {
        self.with_format(|writer| writer.write_all(content.as_bytes()))
    }

    /// Write formatted text into this formatter
    pub fn write_fmt(&mut self, content: fmt::Arguments) -> io::Result<()> {
        self.with_format(|writer| writer.write_fmt(content))
    }
}

/// Formatting trait for types to be displayed as markup, the `rome_console`
/// equivalent to [std::fmt::Display]
///
/// # Example
/// Implementing `Display` on a custom struct
/// ```
/// use std::io;
/// use rome_console::{fmt::{Display, Formatter}, markup};
///
/// struct Warning(String);
///
/// impl Display for Warning {
///     fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
///         fmt.write_markup(markup! {
///             <Warn>{self.0}</Warn>
///         })
///     }
/// }
///
/// let warning = Warning(String::from("content"));
/// markup! {
///     <Emphasis>{warning}</Emphasis>
/// };
/// ```
pub trait Display {
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()>;
}

// Blanket implementations of Display for reference types
impl<'a, T> Display for &'a T
where
    T: Display + ?Sized,
{
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        T::fmt(self, fmt)
    }
}

impl<'a, T> Display for Cow<'a, T>
where
    T: Display + ToOwned + ?Sized,
{
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        T::fmt(self, fmt)
    }
}

// Simple implementations of Display calling through to write_str for types
// that implement Deref<str>
impl Display for str {
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        fmt.write_str(self)
    }
}

impl Display for String {
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        fmt.write_str(self)
    }
}

// Implement Display for Markup and Rust format Arguments
impl<'a> Display for Markup<'a> {
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        fmt.write_markup(*self)
    }
}

impl<'a> Display for std::fmt::Arguments<'a> {
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        fmt.write_fmt(*self)
    }
}

/// Implement [Display] for types that implement [std::fmt::Display] by calling
/// through to [Formatter::write_fmt]
macro_rules! impl_std_display {
    ($ty:ty) => {
        impl Display for $ty {
            fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
                write!(fmt, "{self}")
            }
        }
    };
}

impl_std_display!(char);
impl_std_display!(i8);
impl_std_display!(i16);
impl_std_display!(i32);
impl_std_display!(i64);
impl_std_display!(i128);
impl_std_display!(isize);
impl_std_display!(u8);
impl_std_display!(u16);
impl_std_display!(u32);
impl_std_display!(u64);
impl_std_display!(u128);
impl_std_display!(usize);

impl Display for Duration {
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        use crate as rome_console;

        let secs = self.as_secs();
        if secs > 1 {
            return fmt.write_markup(markup! {
                {secs}<Dim>"s"</Dim>
            });
        }

        let millis = self.as_millis();
        if millis > 1 {
            return fmt.write_markup(markup! {
                {millis}<Dim>"ms"</Dim>
            });
        }

        let micros = self.as_micros();
        if micros > 1 {
            return fmt.write_markup(markup! {
                {micros}<Dim>"Âµs"</Dim>
            });
        }

        let nanos = self.as_nanos();
        fmt.write_markup(markup! {
            {nanos}<Dim>"ns"</Dim>
        })
    }
}
