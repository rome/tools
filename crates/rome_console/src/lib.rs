use atty::Stream;
use std::io;
use std::io::{Read, Stdin, Write};
use std::panic::RefUnwindSafe;
use termcolor::{ColorChoice, StandardStream};
use write::Termcolor;

pub mod fmt;
mod markup;
mod write;

pub use self::markup::{Markup, MarkupBuf, MarkupElement, MarkupNode};
use crate::fmt::Formatter;
pub use rome_markup::markup;

/// Determines the "ouput stream" a message should get printed to
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum LogLevel {
    /// Print the message to the `Error` stream of the console, for instance
    /// "stderr" for the [EnvConsole]
    Error,
    /// Print the message to the `Log` stream of the console, for instance
    /// "stdout" for the [EnvConsole]
    Log,
}

/// Generic abstraction over printing markup and diagnostics to an output,
/// which can be a terminal, a file, a memory buffer ...
pub trait Console: Send + Sync + RefUnwindSafe {
    /// Prints a message (formatted using [markup!]) to the console
    fn print(&mut self, level: LogLevel, args: Markup);

    /// It reads from a source, and if this source contains something, it's converted into a [String]
    fn read(&mut self) -> Option<String>;
}

/// Extension trait for [Console] providing convenience printing methods
pub trait ConsoleExt: Console {
    /// Prints a piece of markup with level [LogLevel::Error]
    fn error(&mut self, args: Markup);

    /// Prints a piece of markup with level [LogLevel::Log]
    fn log(&mut self, args: Markup);
}

impl<T: Console + ?Sized> ConsoleExt for T {
    fn error(&mut self, args: Markup) {
        self.print(LogLevel::Error, args);
    }

    fn log(&mut self, args: Markup) {
        self.print(LogLevel::Log, args);
    }
}

/// Implementation of [Console] printing messages to the standard output and standard error
pub struct EnvConsole {
    /// Channel to print messages
    out: StandardStream,
    /// Channel to print errors
    err: StandardStream,
    /// Channel to read arbitrary input
    r#in: Stdin,
}

impl EnvConsole {
    pub fn new(no_colors: bool) -> Self {
        let out_mode = if no_colors || !atty::is(atty::Stream::Stdout) {
            ColorChoice::Never
        } else {
            ColorChoice::Auto
        };
        let err_mode = if no_colors || !atty::is(atty::Stream::Stderr) {
            ColorChoice::Never
        } else {
            ColorChoice::Auto
        };

        Self {
            out: StandardStream::stdout(out_mode),
            err: StandardStream::stderr(err_mode),
            r#in: io::stdin(),
        }
    }
}

impl Default for EnvConsole {
    fn default() -> Self {
        Self::new(false)
    }
}

impl Console for EnvConsole {
    fn print(&mut self, level: LogLevel, args: Markup) {
        let mut out = match level {
            LogLevel::Error => self.err.lock(),
            LogLevel::Log => self.out.lock(),
        };

        fmt::Formatter::new(&mut Termcolor(&mut out))
            .write_markup(args)
            .unwrap();

        writeln!(out).unwrap();
    }

    fn read(&mut self) -> Option<String> {
        // Here we check if stdin is redirected. If not, we bail.
        //
        // Doing this check allows us to pipe stdin to rome, without expecting
        // user content when we call `read_to_string`
        if atty::is(Stream::Stdin) {
            return None;
        }
        let mut handle = self.r#in.lock();
        let mut buffer = String::new();
        let result = handle.read_to_string(&mut buffer);
        // Skipping the error for now
        if result.is_ok() {
            Some(buffer)
        } else {
            None
        }
    }
}

/// Implementation of [Console] storing all printed messages to a memory buffer
#[derive(Default, Debug)]
pub struct BufferConsole {
    pub out_buffer: Vec<Message>,
    pub in_buffer: Vec<String>,
}

/// Individual message entry printed to a [BufferConsole]
#[derive(Debug)]
pub struct Message {
    pub level: LogLevel,
    pub content: MarkupBuf,
}

impl Console for BufferConsole {
    fn print(&mut self, level: LogLevel, args: Markup) {
        self.out_buffer.push(Message {
            level,
            content: args.to_owned(),
        });
    }
    fn read(&mut self) -> Option<String> {
        if self.in_buffer.is_empty() {
            None
        } else {
            // for the time being we simple return the first message, as we don't
            // particular use case for multiple prompts
            Some(self.in_buffer[0].clone())
        }
    }
}

/// A horizontal line with the given print width
pub struct HorizontalLine {
    width: usize,
}

impl HorizontalLine {
    pub fn new(width: usize) -> Self {
        Self { width }
    }
}

impl fmt::Display for HorizontalLine {
    fn fmt(&self, fmt: &mut Formatter) -> io::Result<()> {
        fmt.write_str(&"\u{2501}".repeat(self.width))
    }
}
