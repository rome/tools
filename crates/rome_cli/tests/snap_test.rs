use rome_console::fmt::{Formatter, Termcolor};
use rome_console::{markup, BufferConsole, Markup};
use rome_diagnostics::termcolor::NoColor;
use rome_fs::{FileSystemExt, MemoryFileSystem};
use std::collections::HashMap;
use std::fmt::Write as _;
use std::path::PathBuf;

#[derive(Default)]
struct InMessages {
    stdin: Option<String>,
}

#[derive(Default)]
struct CliSnapshot {
    /// input messages, coming from different sources
    pub in_messages: InMessages,
    /// the configuration, if set
    pub configuration: Option<String>,
    /// file name -> content
    pub files: HashMap<String, String>,
    /// messages written in console
    pub messages: Vec<String>,
}

impl CliSnapshot {
    fn emit_content_snapshot(&self) -> String {
        let mut content = String::new();

        if let Some(configuration) = &self.configuration {
            content.push_str("## `rome.json`\n\n");
            content.push_str("```json");
            content.push('\n');
            content.push_str(configuration);
            content.push('\n');
            content.push_str("```");
            content.push_str("\n\n")
        }

        for (name, file_content) in &self.files {
            if !name.starts_with("rome.json") {
                let extension = name.split('.').last().unwrap();

                let _ = write!(content, "## `{name}`\n\n");
                let _ = write!(content, "```{extension}");
                content.push('\n');
                content.push_str(&**file_content);
                content.push('\n');
                content.push_str("```");
                content.push_str("\n\n")
            }
        }

        if let Some(stdin) = &self.in_messages.stdin {
            content.push_str("# Input messages\n\n");
            content.push_str("```block");
            content.push('\n');
            content.push_str(stdin);
            content.push('\n');
            content.push_str("```");
            content.push_str("\n\n")
        }

        if !self.messages.is_empty() {
            content.push_str("# Emitted Messages\n\n");

            for message in &self.messages {
                let message_content = &**message;
                // There are some logs that print the timing, and we can't snapshot that message
                // otherwise at each run we invalid the previous snapshot.
                //
                // This is a workaround, and it might not work for all cases.
                if !message_content.contains("files") {
                    content.push_str("```block");
                    content.push('\n');
                    content.push_str(message_content);
                    content.push('\n');
                    content.push_str("```");
                    content.push_str("\n\n")
                }
            }
        }

        content
    }
}

pub fn markup_to_string(markup: Markup) -> String {
    let mut buffer = Vec::new();
    let mut write = Termcolor(NoColor::new(&mut buffer));
    let mut fmt = Formatter::new(&mut write);
    fmt.write_markup(markup).unwrap();

    String::from_utf8(buffer).unwrap()
}

/// Function used to snapshot a session test of the a CLI run.
pub fn assert_cli_snapshot(test_name: &str, fs: MemoryFileSystem, console: BufferConsole) {
    let mut cli_snapshot = CliSnapshot::default();
    let config_path = PathBuf::from("rome.json");
    let configuration = fs.read(&config_path).ok();
    if let Some(mut configuration) = configuration {
        let mut buffer = String::new();
        if configuration.read_to_string(&mut buffer).is_ok() {
            cli_snapshot.configuration = Some(buffer);
        }
    }

    for (file, entry) in fs.files() {
        let content = entry.lock();
        let content = std::str::from_utf8(content.as_slice()).unwrap();

        cli_snapshot
            .files
            .insert(file.to_str().unwrap().to_string(), String::from(content));
    }

    let in_buffer = &console.in_buffer;
    for (index, message) in in_buffer.iter().enumerate() {
        if index == 0 {
            cli_snapshot.in_messages.stdin = Some(message.to_string());
        }
    }

    for message in &console.out_buffer {
        let content = markup_to_string(markup! {
            {message.content}
        });
        cli_snapshot.messages.push(content)
    }

    let content = cli_snapshot.emit_content_snapshot();

    insta::with_settings!({
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_snapshot!(test_name, content);
    });
}
