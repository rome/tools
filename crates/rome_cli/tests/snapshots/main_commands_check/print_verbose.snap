---
source: crates/rome_cli/tests/snap_test.rs
expression: content
---
## `check.js`

```js
for(;true;);

```

# Termination Message

```block
internalError/io ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × Some errors were emitted while running checks
  


```

# Emitted Messages

```block
check.js:1:1 lint/style/useWhile  FIXABLE  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × Use while loops instead of for loops.
  
  > 1 │ for(;true;);
      │ ^^^^^^^^^^^
    2 │ 
  
  i Suggested fix: Use a while loop
  
    1   │ - for(;true;);
      1 │ + while·(true);
    2 2 │   
  

```

```block
Checked 1 file(s) in <TIME>
```


